MainMovement.py

```diff
@@ -0,0 +1,227 @@
+                print("leidis seriali")
+                break
+        if self.ser == None:
+            raise Exception("Could not establish serial connection")
+
+    def close(self):
+        print("sulgesin seriali")
+        self.ser.close()
+
+    def sendCommand(self, thrower_speed, speed1, speed2, speed3):
+        self.ser.write(struct.pack('<hhhHBH', speed1, speed2, speed3, thrower_speed, self.disable_failsafe, 0xAAAA))
+        # reads as many bytes as there are to read
+        self.ser.read(self.ser.in_waiting)
+
+    def move(self, x_speed, y_speed, robotAngularVelocity, thrower_speed):
+        robotSpeed = math.sqrt(x_speed * x_speed + y_speed * y_speed)
+        robotDirectionAngle = math.atan2(y_speed, x_speed)
+        wheelSpeeds = [robotSpeed * math.cos(robotDirectionAngle - angle) +
+                       self.wheelDistanceFromCenter * robotAngularVelocity for angle in self.wheelAngles]
+        wheelSpeedsInMainboardUnits = [round(speed*self.wheelSpeedToMainboardUnits) for speed in wheelSpeeds]
+        # thrower_speed = 0 #temporary
+        # print(wheelSpeedsInMainboardUnits)
+        self.sendCommand(thrower_speed, wheelSpeedsInMainboardUnits[0], wheelSpeedsInMainboardUnits[1], wheelSpeedsInMainboardUnits[2])
+
+
+class States(Enum):
+    spin = 0
+    drive = 1
+    orbit = 2
+    throw = 3
+    controller = 4
+    enemy_basket_spin = 5
+    enemy_basket_approach = 6
+
+
+class StateMachine():
+    def __init__(self, motion, frame_width, frame_height, throwtime):
+        self.motion = motion
+        self.frame_centre_x = frame_width/2
+        self.frame_height = frame_height
+        self.throwtime = throwtime
+
+    def spin(self, processedData,timeoutframes):
+        if timeoutframes > 500:
+            return States.enemy_basket_spin
+
+        # if it finds more than 1 ball to choose from, it goes to the drive function
+        if len(processedData.balls) > 0:
+            #print(processedData.balls)
+            return States.drive
+        else:
+            # otherwise it spins until it finds some
+            self.motion.move(0, 0, 1, 0)
+        return States.spin
+
+    def drive(self, processedData):
+        if len(processedData.balls) > 0:
+            ball = processedData.balls[-1]
+            #print("distance: {}".format(ball))
+            xmiddif = (self.frame_centre_x - (ball.x)) * 0.0075              #approcing ball relative to the ball's x and y coordinate
+            yfor = (self.frame_height*0.75 - (ball.distance)) * 0.005
+            #print("x: ", xmiddif, ", y: ", yfor)
+            self.motion.move(0, yfor, xmiddif, 0)
+            if ball.distance >= self.frame_height*0.73:
+                # if it is close enough to a ball, it should start orbiting around it to find the basket:
+                return States.orbit
+        # if it loses the ball for some reason (opponent takes it), it starts spinning to look for a new one:
+        else:
```

Comment:
I suggest removing the else clause and returning the spin state on line 107. Then I would move the `return States.drive` after the if expression `if ball.distance >= self.frame_height*0.73:`